File: ios/Classes/StyleController.swift
@_spi(Experimental) import MapboxMaps
import Foundation
import Flutter

final class StyleController: StyleManager {
    private static let errorCode = "0"

    private let styleManager: MapboxMaps.StyleManager

    init(styleManager: MapboxMaps.StyleManager) {
        self.styleManager = styleManager
    }

    func getStyleURI(completion: @escaping (Result<String, Error>) -> Void) {
        completion(.success(styleManager.styleURI?.rawValue ?? ""))
    }

    func setStyleURI(uri: String, completion: @escaping (Result<Void, Error>) -> Void) {
        guard let styleURI = StyleURI(rawValue: uri) else {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: "Invalid style uri", details: nil)))
            return
        }
        styleManager.load(mapStyle: MapStyle(uri: styleURI)) { error in
            guard let error else {
                completion(.success(()))
                return
            }
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func getStyleJSON(completion: @escaping (Result<String, Error>) -> Void) {
        completion(.success(styleManager.styleJSON))
    }

    func setStyleJSON(json: String, completion: @escaping (Result<Void, Error>) -> Void) {
        styleManager.load(mapStyle: MapStyle(json: json)) { error in
            guard let error else {
                completion(.success(()))
                return
            }

            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func getStyleDefaultCamera(completion: @escaping (Result<CameraOptions, Error>) -> Void) {
        let camera = styleManager.styleDefaultCamera
        completion(.success(camera.toFLTCameraOptions()))
    }

    func getStyleTransition(completion: @escaping (Result<TransitionOptions, Error>) -> Void) {
        let transition = styleManager.styleTransition
        completion(.success(transition.toFLTTransitionOptions()))

    }

    func setStyleTransition(transitionOptions: TransitionOptions, completion: @escaping (Result<Void, Error>) -> Void) {
        styleManager.styleTransition = transitionOptions.toTransitionOptions()
        completion(.success(()))
    }

    func addStyleLayer(properties: String, layerPosition: LayerPosition?, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            let layerProperties: [String: Any] = convertStringToDictionary(properties: properties)
            try styleManager.addLayer(with: layerProperties, layerPosition: layerPosition?.toLayerPosition())
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func addPersistentStyleLayer(properties: String, layerPosition: LayerPosition?, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            try styleManager.addPersistentLayer(
                with: convertStringToDictionary(properties: properties),
                layerPosition: layerPosition?.toLayerPosition()
            )
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func isStyleLayerPersistent(layerId: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        do {
            let isPersistent = try styleManager.isPersistentLayer(id: layerId)
            completion(.success(isPersistent))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func removeStyleLayer(layerId: String, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            try styleManager.removeLayer(withId: layerId)
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func moveStyleLayer(layerId: String, layerPosition: LayerPosition?, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            if layerPosition != nil {
                try styleManager.moveLayer(withId: layerId, to: layerPosition!.toLayerPosition())
            } else {
                try styleManager.moveLayer(withId: layerId, to: .default)
            }
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func styleLayerExists(layerId: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        let exists = styleManager.layerExists(withId: layerId)
        completion(.success(exists))
    }

    func getStyleLayers(completion: @escaping (Result<[StyleObjectInfo?], Error>) -> Void) {
        let layerInfos = styleManager.allLayerIdentifiers.map {
            StyleObjectInfo(id: $0.id, type: $0.type.rawValue)
        }
        completion(.success(layerInfos))
    }

    func getStyleLayerProperty(layerId: String, property: String, completion: @escaping (Result<StylePropertyValue, Error>) -> Void) {
        let layerProperty = styleManager.layerProperty(for: layerId, property: property)
        completion(.success(layerProperty.toFLTStylePropertyValue(property: property)))
    }

    func setStyleLayerProperty(layerId: String, property: String, value: Any, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            var mappedValue = value
            if let stringValue = value as? String {
                if stringValue.hasPrefix("[") || stringValue.hasPrefix("{") {
                    if let expressionData = stringValue.data(using: .utf8) {
                        let expJSONObject = try JSONSerialization.jsonObject(with: expressionData, options: [])
                        mappedValue = expJSONObject
                    }
                }
            }
            try styleManager.setLayerProperty(for: layerId, property: property, value: mappedValue)
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func getStyleLayerProperties(layerId: String, completion: @escaping (Result<String, Error>) -> Void) {
        do {
            let properties = try styleManager.layerProperties(for: layerId)
            completion(.success(convertDictionaryToString(dict: properties)))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func setStyleLayerProperties(layerId: String, properties: String, completion: @escaping (Result<Void, Error>) -> Void) {
        let data = properties.data(using: String.Encoding.utf8)!
        let jsonObject = try? JSONSerialization.jsonObject(with: data, options: [])
        do {
            try styleManager.setLayerProperties(for: layerId, properties: jsonObject as? [String: Any] ?? [:])
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func addStyleSource(sourceId: String, properties: String, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            try styleManager.addSource(withId: sourceId,
                                          properties: convertStringToDictionary(properties: properties))
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func getStyleSourceProperty(sourceId: String, property: String, completion: @escaping (Result<StylePropertyValue, Error>) -> Void) {
        let sourceProperty = styleManager.sourceProperty(for: sourceId, property: property)
        completion(.success(sourceProperty.toFLTStylePropertyValue(property: property)))
    }

    func setStyleSourceProperty(sourceId: String, property: String, value: Any, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            try styleManager.setSourceProperty(for: sourceId, property: property, value: value)
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func getStyleSourceProperties(sourceId: String, completion: @escaping (Result<String, Error>) -> Void) {
        do {
            let properties = try styleManager.sourceProperties(for: sourceId)
            completion(.success(convertDictionaryToString(dict: properties)))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func setStyleSourceProperties(sourceId: String, properties: String, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            try styleManager.setSourceProperties(for: sourceId, properties: convertStringToDictionary(properties: properties))
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func addGeoJSONSourceFeatures(sourceId: String, dataId: String, features: [Feature], completion: @escaping (Result<Void, Error>) -> Void) {
        styleManager.addGeoJSONSourceFeatures(forSourceId: sourceId, features: features, dataId: dataId)
        completion(.success(()))
    }

    func updateGeoJSONSourceFeatures(sourceId: String, dataId: String, features: [Feature], completion: @escaping (Result<Void, Error>) -> Void) {
        styleManager.updateGeoJSONSourceFeatures(forSourceId: sourceId, features: features, dataId: dataId)
        completion(.success(()))
    }

    func removeGeoJSONSourceFeatures(sourceId: String, dataId: String, featureIds: [String], completion: @escaping (Result<Void, Error>) -> Void) {
        styleManager.removeGeoJSONSourceFeatures(forSourceId: sourceId, featureIds: featureIds, dataId: dataId)
        completion(.success(()))
    }

    func updateStyleImageSourceImage(sourceId: String, image: MbxImage, completion: @escaping (Result<Void, Error>) -> Void) {
        guard let image = UIImage(data: image.data.data, scale: UIScreen.main.scale) else {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: "Could not initialize the image from the specified data.", details: nil)))
            return
        }
        do {
            try styleManager.updateImageSource(withId: sourceId, image: image)
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func removeStyleSource(sourceId: String, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            try styleManager.removeSource(withId: sourceId)
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func styleSourceExists(sourceId: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        let exists = styleManager.sourceExists(withId: sourceId)
        completion(.success(exists))
    }

    func getStyleSources(completion: @escaping (Result<[StyleObjectInfo?], Error>) -> Void) {
        let sourcesInfos = styleManager.allSourceIdentifiers.map {
            StyleObjectInfo(id: $0.id, type: $0.type.rawValue)
        }
        completion(.success(sourcesInfos))
    }

    func setStyleTerrain(properties: String, completion: @escaping (Result<Void, Error>) -> Void) {
        let data = properties.data(using: String.Encoding.utf8)!
        let jsonObject = try? JSONSerialization.jsonObject(with: data, options: [])
        do {
            try styleManager.setTerrain(properties: jsonObject as? [String: Any] ?? [:])
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func getStyleTerrainProperty(property: String, completion: @escaping (Result<StylePropertyValue, Error>) -> Void) {
        let terrainProperty: MapboxMaps.StylePropertyValue = styleManager.terrainProperty(property)
        completion(.success(terrainProperty.toFLTStylePropertyValue(property: property)))
    }

    func setStyleTerrainProperty(property: String, value: Any, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            try styleManager.setTerrainProperty(property, value: value)
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func getStyleImage(imageId: String, completion: @escaping (Result<MbxImage?, Error>) -> Void) {
        guard let image = styleManager.image(withId: imageId) else {
            completion(.success(nil))
            return
        }

        let data = FlutterStandardTypedData(bytes: image.pngData()!)

        completion(.success(MbxImage(width: Int64(image.size.width * image.scale),
                                    height: Int64(image.size.height * image.scale),
                                    data: data)))
    }

    func addStyleImage(imageId: String, scale: Double, image: MbxImage, sdf: Bool, stretchX: [ImageStretches?], stretchY: [ImageStretches?], content: ImageContent?, completion: @escaping (Result<Void, Error>) -> Void) {
        guard let image = UIImage(data: image.data.data, scale: scale) else {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: "Could not initialize the image from the specified data.", details: nil)))
            return
        }
        var imageContent: MapboxMaps.ImageContent?
        if let content {
            imageContent = MapboxMaps.ImageContent(left: Float(content.left),
                                        top: Float(content.top),
                                        right: Float(content.right),
                                        bottom: Float(content.bottom))
        }
        do {
            try styleManager.addImage(image,
                                         id: imageId,
                                         sdf: sdf,
                                      stretchX: stretchX.compactMap {$0 }.map {
                MapboxMaps.ImageStretches(first: Float($0.first), second: Float($0.second))

            }, stretchY: stretchY.compactMap {$0 }.map {
                MapboxMaps.ImageStretches(first: Float($0.first), second: Float($0.second))},
                                         content: imageContent)
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func removeStyleImage(imageId: String, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            try styleManager.removeImage(withId: imageId)
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func hasStyleImage(imageId: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        let imageExists = styleManager.imageExists(withId: imageId)
        completion(.success(imageExists))
    }

    func invalidateStyleCustomGeometrySourceTile(sourceId: String, tileId: CanonicalTileID, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            try styleManager.invalidateCustomGeometrySourceTile(forSourceId: sourceId,
                                                                   tileId: tileId.toCanonicalTileID())
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func invalidateStyleCustomGeometrySourceRegion(sourceId: String, bounds: CoordinateBounds, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            try styleManager.invalidateCustomGeometrySourceRegion(forSourceId: sourceId,
                                                                     bounds: bounds.toCoordinateBounds())
            completion(.success(()))
        } catch {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: error.localizedDescription, details: nil)))
        }
    }

    func isStyleLoaded(completion: @escaping (Result<Bool, Error>) -> Void) {
        completion(.success(styleManager.isStyleLoaded))
    }

    func localizeLabels(locale: String, layerIds: [String]?, completion: @escaping (Result<Void, Error>) -> Void) {
        try! styleManager.localizeLabels(into: Locale(identifier: locale), forLayerIds: layerIds)
        completion(.success(()))
    }

    // MARK: Style Imports

    func getStyleImports() throws -> [StyleObjectInfo?] {
        styleManager.styleImports.map { StyleObjectInfo(id: $0.id, type: $0.type) }
    }

    func removeStyleImport(importId: String) throws {
        do {
            try styleManager.removeStyleImport(withId: importId)
        } catch let styleError {
            throw FlutterError(code: StyleController.errorCode, message: styleError.localizedDescription, details: nil)
        }
    }

    func getStyleImportSchema(importId: String) throws -> Any {
        do {
            return try styleManager.getStyleImportSchema(for: importId)
        } catch let styleError {
            throw FlutterError(code: StyleController.errorCode, message: styleError.localizedDescription, details: nil)
        }
    }

    func getStyleImportConfigProperties(importId: String) throws -> [String: StylePropertyValue] {
        do {
            let styleImportsConfig = try styleManager.getStyleImportConfigProperties(for: importId)
            return styleImportsConfig.reduce(into: [:]) { partialResult, pair in
                let (key, value) = pair
                partialResult[key] = value.toFLTStylePropertyValue(property: key)
            }
        } catch let styleError {
            throw FlutterError(code: StyleController.errorCode, message: styleError.localizedDescription, details: nil)
        }
    }

    func getStyleImportConfigProperty(importId: String, config: String) throws -> StylePropertyValue {
        do {
            let value = try styleManager.getStyleImportConfigProperty(for: importId, config: config)
            return value.toFLTStylePropertyValue(property: config)
        } catch let styleError {
            throw FlutterError(code: StyleController.errorCode, message: styleError.localizedDescription, details: nil)
        }
    }

    func setStyleImportConfigProperties(importId: String, configs: [String: Any]) throws {
        do {
            try styleManager.setStyleImportConfigProperties(for: importId, configs: configs)
        } catch let styleError {
            throw FlutterError(code: StyleController.errorCode, message: styleError.localizedDescription, details: nil)
        }
    }

    func setStyleImportConfigProperty(importId: String, config: String, value: Any) throws {
        do {
            try styleManager.setStyleImportConfigProperty(for: importId, config: config, value: value)
        } catch let styleError {
            throw FlutterError(code: StyleController.errorCode, message: styleError.localizedDescription, details: nil)
        }
    }

    // MARK: Style Lights

    func getStyleLights() throws -> [StyleObjectInfo?] {
        styleManager.allLightIdentifiers.map {
            StyleObjectInfo(id: $0.id, type: $0.type.rawValue)
        }
    }

    func setLight(flatLight: FlatLight) throws {
        do {
            try styleManager.setLights(MapboxMaps.FlatLight(flatLight))
        } catch let styleError {
            throw FlutterError(code: StyleController.errorCode, message: styleError.localizedDescription, details: nil)
        }
    }

    func setLights(ambientLight: AmbientLight, directionalLight: DirectionalLight) throws {
        do {
            try styleManager.setLights(ambient: MapboxMaps.AmbientLight(ambientLight), directional: MapboxMaps.DirectionalLight(directionalLight))
        } catch let styleError {
            throw FlutterError(code: StyleController.errorCode, message: styleError.localizedDescription, details: nil)
        }
    }

    func getStyleLightProperty(id: String, property: String, completion: @escaping (Result<StylePropertyValue, Error>) -> Void) {
        let value = styleManager.lightProperty(for: id, property: property)
        var kind = MapboxMaps.StylePropertyValueKind.constant
        // FIXME: Remove workaround to get property kind one MapboxMaps iOS SDK updates.
        if property.hasSuffix("transition") {
            kind = .transition
        }
        completion(.success(MapboxMaps.StylePropertyValue(value: value, kind: kind).toFLTStylePropertyValue(property: property)))
    }

    func setStyleLightProperty(id: String, property: String, value: Any, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            try styleManager.setLightProperty(for: id, property: property, value: value)
            completion(.success(()))
        } catch let styleError {
            completion(.failure(FlutterError(code: StyleController.errorCode, message: styleError.localizedDescription, details: nil)))
        }
    }

    // MARK: Style Projection

    func getProjection() throws -> StyleProjection? {
        return styleManager.projection?.toFLTStyleProjection()
    }

    func setProjection(projection: StyleProjection) throws {
        do {
            try styleManager.setProjection(MapboxMaps.StyleProjection(name: MapboxMaps.StyleProjectionName(projection.name)))
        } catch let styleError {
            throw FlutterError(code: StyleController.errorCode, message: styleError.localizedDescription, details: nil)
        }
    }

    // MARK: Model
    func addStyleModel(modelId: String, modelUri: String, completion: @escaping (Result<Void, any Error>) -> Void) {
        completion(.init(code: StyleController.errorCode, catchingFlutter: {
            try styleManager.addStyleModel(modelId: modelId, modelUri: modelUri)
        }))
    }

    func removeStyleModel(modelId: String, completion: @escaping (Result<Void, any Error>) -> Void) {
        completion(.init(code: StyleController.errorCode, catchingFlutter: {
            try styleManager.removeStyleModel(modelId: modelId)
        }))
    }

}

File: mapbox_maps_flutter_mobile/lib/src/controllers/style_controller.dart
part of 'package:mapbox_maps_flutter_mobile/mapbox_maps_flutter_mobile.dart';

/// Mobile [StylePlatformInterface] implementation, backed by the
/// pigeon-generated [StyleManager]. Forwards every call as-is, except the
/// image methods, which convert the public [StyleImage] to and from the
/// pigeon [StyleImageWire] crossed over the platform channel.
class StyleController implements StylePlatformInterface {
  StyleController(this._api);

  final StyleManager _api;

  // ===== Style loading =====

  @override
  Future<void> setStyleURI(String uri) => _api.setStyleURI(uri);

  @override
  Future<String> getStyleURI() => _api.getStyleURI();

  @override
  Future<void> setStyleJSON(String json) => _api.setStyleJSON(json);

  @override
  Future<String> getStyleJSON() => _api.getStyleJSON();

  // ===== Default camera & transition =====

  @override
  Future<CameraOptions> getStyleDefaultCamera() => _api.getStyleDefaultCamera();

  @override
  Future<TransitionOptions> getStyleTransition() => _api.getStyleTransition();

  @override
  Future<void> setStyleTransition(TransitionOptions transitionOptions) =>
      _api.setStyleTransition(transitionOptions);

  // ===== Style imports =====

  @override
  Future<void> addStyleImportFromJSON(
    String importId,
    String json, {
    Map<String, Object>? config,
    ImportPosition? importPosition,
  }) => _api.addStyleImportFromJSON(
    importId,
    json,
    config: config,
    importPosition: importPosition,
  );

  @override
  Future<void> addStyleImportFromURI(
    String importId,
    String uri, {
    Map<String, Object>? config,
    ImportPosition? importPosition,
  }) => _api.addStyleImportFromURI(
    importId,
    uri,
    config: config,
    importPosition: importPosition,
  );

  @override
  Future<void> updateStyleImportWithJSON(
    String importId,
    String json, {
    Map<String, Object>? config,
  }) => _api.updateStyleImportWithJSON(importId, json, config: config);

  @override
  Future<void> updateStyleImportWithURI(
    String importId,
    String uri, {
    Map<String, Object>? config,
  }) => _api.updateStyleImportWithURI(importId, uri, config: config);

  @override
  Future<void> moveStyleImport(
    String importId,
    ImportPosition? importPosition,
  ) => _api.moveStyleImport(importId, importPosition);

  @override
  Future<List<StyleObjectInfo?>> getStyleImports() => _api.getStyleImports();

  @override
  Future<void> removeStyleImport(String importId) =>
      _api.removeStyleImport(importId);

  @override
  Future<Object> getStyleImportSchema(String importId) =>
      _api.getStyleImportSchema(importId);

  @override
  Future<Map<String, StylePropertyValue>> getStyleImportConfigProperties(
    String importId,
  ) => _api.getStyleImportConfigProperties(importId);

  @override
  Future<StylePropertyValue> getStyleImportConfigProperty(
    String importId,
    String config,
  ) => _api.getStyleImportConfigProperty(importId, config);

  // ===== Layers =====

  @override
  Future<void> addStyleLayer(String properties, LayerPosition? layerPosition) =>
      _api.addStyleLayer(properties, layerPosition);

  @override
  Future<void> addPersistentStyleLayer(
    String properties,
    LayerPosition? layerPosition,
  ) => _api.addPersistentStyleLayer(properties, layerPosition);

  @override
  Future<bool> isStyleLayerPersistent(String layerId) =>
      _api.isStyleLayerPersistent(layerId);

  @override
  Future<StylePropertyValue> getStyleLayerProperty(
    String layerId,
    String property,
  ) => _api.getStyleLayerProperty(layerId, property);

  @override
  Future<List<StyleObjectInfo?>> getStyleLayers() => _api.getStyleLayers();

  @override
  Future<bool> styleLayerExists(String layerId) =>
      _api.styleLayerExists(layerId);

  @override
  Future<void> removeStyleLayer(String layerId) =>
      _api.removeStyleLayer(layerId);

  @override
  Future<void> moveStyleLayer(String layerId, LayerPosition? layerPosition) =>
      _api.moveStyleLayer(layerId, layerPosition);

  // ===== Sources =====

  @override
  Future<void> addStyleSource(String sourceId, String properties) =>
      _api.addStyleSource(sourceId, properties);

  @override
  Future<List<StyleObjectInfo?>> getStyleSources() => _api.getStyleSources();

  @override
  Future<bool> styleSourceExists(String sourceId) =>
      _api.styleSourceExists(sourceId);

  @override
  Future<void> removeStyleSource(String sourceId) =>
      _api.removeStyleSource(sourceId);

  // ===== GeoJSON source partial updates =====

  @override
  Future<void> addGeoJSONSourceFeatures(
    String sourceId,
    String dataId,
    List<Feature> features,
  ) => _api.addGeoJSONSourceFeatures(sourceId, dataId, features);

  @override
  Future<void> updateGeoJSONSourceFeatures(
    String sourceId,
    String dataId,
    List<Feature> features,
  ) => _api.updateGeoJSONSourceFeatures(sourceId, dataId, features);

  @override
  Future<void> removeGeoJSONSourceFeatures(
    String sourceId,
    String dataId,
    List<String> featureIds,
  ) => _api.removeGeoJSONSourceFeatures(sourceId, dataId, featureIds);

  // ===== Images =====

  @override
  Future<bool> hasStyleImage(String imageId) => _api.hasStyleImage(imageId);

  @override
  Future<void> addStyleImage(
    String imageId,
    double scale,
    StyleImage image,
    bool sdf,
    List<ImageStretches?> stretchX,
    List<ImageStretches?> stretchY,
    ImageContent? content,
  ) => _api.addStyleImage(
    imageId,
    scale,
    image.toWire(),
    sdf,
    stretchX,
    stretchY,
    content,
  );

  @override
  Future<void> updateStyleImageSourceImage(String sourceId, StyleImage image) =>
      _api.updateStyleImageSourceImage(sourceId, image.toWire());

  @override
  Future<void> removeStyleImage(String imageId) =>
      _api.removeStyleImage(imageId);

  // ===== Models =====

  @override
  Future<void> addStyleModel(String modelId, String modelUri) =>
      _api.addStyleModel(modelId, modelUri);

  @override
  Future<void> removeStyleModel(String modelId) =>
      _api.removeStyleModel(modelId);

  // ===== Lights =====

  @override
  Future<List<StyleObjectInfo?>> getStyleLights() => _api.getStyleLights();

  @override
  Future<void> setLight(FlatLight flatLight) => _api.setLight(flatLight);

  @override
  Future<void> setLights(
    AmbientLight ambientLight,
    DirectionalLight directionalLight,
  ) => _api.setLights(ambientLight, directionalLight);

  @override
  Future<StylePropertyValue> getStyleLightProperty(
    String id,
    String property,
  ) => _api.getStyleLightProperty(id, property);

  @override
  Future<void> setStyleLightProperty(
    String id,
    String property,
    Object value,
  ) => _api.setStyleLightProperty(id, property, value);

  // ===== Terrain property access =====

  @override
  Future<StylePropertyValue> getStyleTerrainProperty(String property) =>
      _api.getStyleTerrainProperty(property);

  @override
  Future<void> setStyleTerrainProperty(String property, Object value) =>
      _api.setStyleTerrainProperty(property, value);

  // ===== Image lookup =====

  @override
  Future<StyleImageRgba?> getStyleImage(String imageId) async =>
      (await _api.getStyleImage(imageId))?.toStyleImageRgba();

  // ===== Custom geometry source invalidation =====

  @override
  Future<void> invalidateStyleCustomGeometrySourceTile(
    String sourceId,
    CanonicalTileID tileId,
  ) => _api.invalidateStyleCustomGeometrySourceTile(sourceId, tileId);

  @override
  Future<void> invalidateStyleCustomGeometrySourceRegion(
    String sourceId,
    CoordinateBounds bounds,
  ) => _api.invalidateStyleCustomGeometrySourceRegion(sourceId, bounds);

  // ===== Style state =====

  @override
  Future<bool> isStyleLoaded() => _api.isStyleLoaded();

  @override
  Future<void> localizeLabels(String locale, List<String>? layerIds) =>
      _api.localizeLabels(locale, layerIds);

  @override
  Future<List<FeaturesetDescriptor>> getFeaturesets() => _api.getFeaturesets();

  // ===== Projection =====

  @override
  Future<StyleProjection?> getProjection() => _api.getProjection();

  @override
  Future<void> setProjection(StyleProjection projection) =>
      _api.setProjection(projection);

  // ===== Low-level property access =====

  @override
  Future<void> setStyleLayerProperty(
    String layerId,
    String property,
    Object value,
  ) => _api.setStyleLayerProperty(layerId, property, value);

  @override
  Future<String> getStyleLayerProperties(String layerId) =>
      _api.getStyleLayerProperties(layerId);

  @override
  Future<void> setStyleLayerProperties(String layerId, String properties) =>
      _api.setStyleLayerProperties(layerId, properties);

  @override
  Future<String> getStyleSourceProperties(String sourceId) =>
      _api.getStyleSourceProperties(sourceId);

  @override
  Future<StylePropertyValue> getStyleSourceProperty(
    String sourceId,
    String property,
  ) => _api.getStyleSourceProperty(sourceId, property);

  @override
  Future<void> setStyleSourceProperty(
    String sourceId,
    String property,
    Object value,
  ) => _api.setStyleSourceProperty(sourceId, property, value);

  @override
  Future<void> setStyleSourceProperties(String sourceId, String properties) =>
      _api.setStyleSourceProperties(sourceId, properties);

  @override
  Future<void> setStyleImportConfigProperty(
    String importId,
    String config,
    Object value,
  ) => _api.setStyleImportConfigProperty(importId, config, value);

  @override
  Future<void> setStyleImportConfigProperties(
    String importId,
    Map<String, Object> configs,
  ) => _api.setStyleImportConfigProperties(importId, configs);

  @override
  Future<void> setStyleTerrain(String properties) =>
      _api.setStyleTerrain(properties);
}
